jOOQ/src/main/java/org/jooq/Grant.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Other licenses:
+ * -----------------------------------------------------------------------------
+ * Commercial licenses for this work are available. These replace the above
+ * ASL 2.0 and offer limited warranties, support, maintenance, and commercial
+ * database integrations.
+ *
+ * For more information, please visit: http://www.jooq.org/licenses
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ */
+package org.jooq;
+
+import java.util.Collection;
+
+/**
+ * @author Timur Shaidullin
+ */
+public interface Grant<R extends Record> extends QueryPart, Query {
+
+    public Grant<R> grant(Privilege privilege);
+
+    public Grant<R> grant(Collection<? extends Privilege> privileges);
+
+    public Grant<R> on(Table<?> table);
+
+    public Grant<R> to(User user);
+
+    public Grant<R> to(Role role);
```

Comment:
The design of this class is following a "builder pattern" / "fluent API", but not jOOQ's DSL API strategy. Please, take inspiration by the various other APIs in order to see how this should be done. Your version would allow bogus statements, like:

```java
grant(a).grant(b).on(t1).on(t2).to(a).grant(c).on(t3);
```

Here's a blog post showing how this API design principle works:
https://blog.jooq.org/2012/01/05/the-java-fluent-api-designer-crash-course/

Reply:
Thank you for an article. It's the awesome way to build DSL API strategy